### PR TITLE
[Messenger] multiple failure transports support

### LINKED ISSUE #34911: [Messenger][Feature] Failed queue per transport
**Description**  
Currently we can only set one failed queue that is available for every transport.

What happens if we want different behaviours? 

Like, one queue that we just want to discard the failed queue and others where we care and want to use the failed queue (a specific one or the same global one)?

**Example**  

```yaml
# config/packages/messenger.yaml
framework:
    # no need to set failed transport globally if I want a specific behaviour per transport.
    messenger:
        transports:
            failed: 'doctrine://default?queue_name=failed'
            failed_important: 'doctrine://default?queue_name=failed_important'
            async:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                failure_transport: failed
                retry_strategy:
                    max_retries: 3
                    delay: 1000
                    multiplier: 2
            async_no_failure_transport:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                retry_strategy:
                    max_retries: 3
                    delay: 1000
                    multiplier: 2
            async_send_specific_failure_queue:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                failure_transport: failed_important
                retry_strategy:
                    max_retries: 3
                    delay: 1000
                    multiplier: 2
```

This feature covers for example cases where you need to make distinguish between:
- Events that are constantly being generated and you don't care if they fail after the number of retries
- Events that you need to be absolutely sure that they need to be executed, and moved to a different failed queue (to be debugged by a specific team related with a specific queue. It could even be a very important queue, and you need to prioritise the debug of those events)

I don't mind to try to implement this behaviour. 
Just wanted to know your feedback.

### PR DESCRIPTION
**Pull request by monteiro**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       | Fix #34911 
| License       | MIT
| Doc PR        | symfony/symfony-docs#13489

## Strategy applied

- Pass a map of transports and failed transports to the `SendFailedMessageToFailureTransportListener`. This way we re-use the same listener.
- Local failed transport has more priority than a global failed transport defined. 

## Configuration example

```yaml
# config/packages/messenger.yaml
framework:
    # no need to set failed transport globally if I want a specific behaviour per transport.
    failure_transport: failed # all transports have this failed transport
    messenger:
        transports:
            failed: 'doctrine://default?queue_name=failed'
            failed_important: 'doctrine://default?queue_name=failed_important'
            async:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                failure_transport: failed # takes precedence over the global defined "failed_transport"
                retry_strategy:
                    max_retries: 3
                    delay: 1000
                    multiplier: 2
            async_no_failure_transport: # it will use the global defined transport if no one is defined.
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                retry_strategy:
                    max_retries: 3
                    delay: 1000
                    multiplier: 2
            async_send_specific_failure_queue:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                failed_transport: failed_important # takes precedence over the global defined "failed_transport"
                retry_strategy:
                    max_retries: 3
                    delay: 1000
                    multiplier: 2
```

You can test this feature easily on a [demo project](https://github.com/monteiro/messenger-multiple-failed-transports-pr34979). Just follow the [README](https://github.com/monteiro/messenger-multiple-failed-transports-pr34979).

**More information on issue #34911** 

## What needs to be done so this can be merged:
- [x] validate strategy 
- [ ] update src/**/CHANGELOG.md files
- [x] update tests to cover all cases
- [x] create doc PR

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (monteiro) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: what can we do with the commands for the retries if we have now multiple failed transports.
Should I create one command per transport with a failed queue defined?

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (sroze) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php]: You should throw an error here. Because you `array_fill` in the wiring, this should be all good 👍

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (sroze) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php]: This is a BC-break. Can you make sure to also allow `SenderInterface`, which would be used by all receivers?

Comment (sroze) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php]: Instead of being an `array`, can you use a service locator instead?

**Review by sroze — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (monteiro) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php]: What happens if there is no global failure transport, what should be the value of the SenderInterface?

`SendFailedMessageToFailureTransportListener.php` is instantiated only once. That means that this class is used by all transports to send messages to their failure transport (that can be a global or a local one defined in each transport).

Should I create one `SendFailedMessageToFailureTransportListener` for each failure transport? I would need an extra argument that tells me the transport name, so I can discard messages of different failed transports.

Comment (monteiro) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php]: `$failureSenders` is an array with the transport name and its failure transport.
The service locator doesn't hold this logic.

**Comment by monteiro** [comment]
@sroze thanks a lot for the review. 

Maybe the best way to remove BC, is actually to do a separated listener, but only when multiple ones are defined. What do you think?

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (monteiro) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php]: I understand now the service locator concept 😄thanks!

Comment (monteiro) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php]: My proposal, which I'm going to do next:

The constructor of `SendFailedMessageToFailureTransportListener`:
`public function __construct(?SenderInterface $globalfailureSender, LoggerInterface $logger = null, ServiceLocator $failureSenders)`

Will guarantee that BC is not broken.

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (monteiro) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php]: I changed the logic now. Maybe throwing an error doesn't make sense now.

**Comment by monteiro** [comment]
@sroze Already addressed all the comments to not break BC.

I have some questions as well regarding the failed commands without breaking BC:
- console.command.messenger_failed_messages_retry
- console.command.messenger_failed_messages_show
- console.command.messenger_failed_messages_remove

What should we do about them. Probably the best idea is to create more commands with different failed _messenger transports_ since now it's only one.

**Comment by monteiro** [comment]
The strategy I'm using for failed commands is:
- Pass a failed transport name to the command that is a required argument.
- Passing a service locator of all failed transports to know which failed transport to consume from.

What's missing:
- Unit tests
- More testing (show and retry working, missing remove command).
- Rebase with latest master

**Comment by monteiro** [comment]
Can someone help me with the review? 

I was very careful in order to not break any command or any class by extending some classes with more constructor parameters or adding more options to commands (the failed commands).

**Comment by giovannialbero1992** [comment]
What is missing in this PR? Only Changelog updates?

**Comment by monteiro** [comment]
@giovannialbero1992 and rebasing now that I noticed there are some conflicting files. 
Maybe @weaverryan can help me unblock and give me some hints on this one.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/console.xml
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/console.xml:122]: This should be moved to the new PHP config file.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml:86]: This should be moved to the new PHP config file.

**Review by fabpot — CHANGES_REQUESTED** [review]
Quick review (I've not had a look at the feature itself).

File: src/Symfony/Component/Messenger/Command/AbstractFailedMessagesCommand.php
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Command/AbstractFailedMessagesCommand.php:39]: The purpose of `$receiverName` now is to be the `$globalFailureReceiverName`. I think we should rename the variable for clarity. The idea would be that, *if* this is set AND no `--failed-transport=` option is passed, then we will use this automatically. If this is NOT set, then we could ask the user interactively which failure receiver they want to use.

For the 2nd and 3rd arguments, I think they should be "merged". What I mean is, the "default failure receiver" (if there is one) will be inside of the `ServiceLocator $failureTransports` (if it's not already, it should be). This means that we only need 1 argument here, not 2.

We also need to worry about backwards-compatibility and triggering deprecation notices. So I would:

A) Only have a 2nd argument (not a 2nd and third)
B) Remove the type-hint from the argument (so that it allows both a `ReceiverInterface` or a `ServiceLocator`
C) Inside the constructor, check to see if the argument is `!instanceof ServiceLocator`. In that case, trigger a deprecation warning that says that they SHOULD now pass a locator of failure transports. Then write some code to make it work both ways - one simple way might be to create your own `ServiceLocator` object if they're using the deprecated code path - maybe something like:

```
$this->failureTransports = new ServiceLocator([$this->receiverName => function() { return $failureTransports; }
```

I'm not sure if that code to instantiate the ServiceLocator is 100% correct. But the idea is to create a mini-container with the ONE receiver to make the deprecated code path work. Then the rest of the code can work the same.

File: src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Command/FailedMessagesShowCommand.php:39]: All of these should probably be `failure-transport`, right? That's what we call it in the config

File: src/Symfony/Component/Messenger/Command/AbstractFailedMessagesCommand.php
Comment (weaverryan) [reply src/Symfony/Component/Messenger/Command/AbstractFailedMessagesCommand.php:48]: Also, for the case of (B) (where we are using the global/default transport because no option was passed), we should print a message telling the user how to use different transports (if there are any):

> To use a different failure transport, re-run the command with --failure-transport=
> Available failure transports are: foo, baz, larry

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (weaverryan) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:38]: I would use the same strategy here that I mentioned for the `AbstractFailedMesagesCommand`: do NOT add a new argument, but instead, remove the type-hint from this first argument, trigger a deprecation warning of the old value is passed, and write code to make the old code still work.

**Review by weaverryan — COMMENTED** [review]
We're on the right path!

**Comment by fabpot** [comment]
@monteiro Do you think you will have time in the coming weeks to finish this PR based on the feedback? Or do you need some help?

**Comment by monteiro** [comment]
@fabpot Absolutely. Already started again yesterday. This week I will already have the comments applied.

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (monteiro) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:38]: @weaverryan  I need to get the transport name from a SenderInterface. How can I get that? I need it, because in the event, we get the receiver name and get from the locator the sender to send to the failed transport.

File: src/Symfony/Component/Messenger/Command/AbstractFailedMessagesCommand.php
Comment (monteiro) [reply src/Symfony/Component/Messenger/Command/AbstractFailedMessagesCommand.php:39]: Already talked with @weaverryan about the solution

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml
Comment (monteiro) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml]: As I have discussed with you, we need two locators. 
One for the listener that needs to know the failed transport from the queue, and another one to retry a specific message on a specific failed transport.

**Comment by monteiro** [comment]
@fabpot @weaverryan This PR is ready to be reviewed. 

I created a symfony project [project](https://github.com/symfony/symfony/pull/34979), so anyone can test, even if you don't have the symfony repository downloaded.

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (monteiro) [inline src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:35]: I will need the `globalFailureSender` here, because `failureSenders` is map where key is the transport name. 
When setting a global `failure_transport` is not associated to a transport name.

I need it, because in https://github.com/symfony/symfony/pull/34979/files#diff-2eca49162e4b0a72c3825b50f160d944R72 we will ask the failure sender from the transport the event was consumed.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:1903]: these should be references, right? We don't normally set a Definition directly on an argument. And should it use `$failureTransportsServiceLocatorId` or `$failureTransportsServiceLocator` - I'm getting lost (this stuff confuses me) in which is which and what should be used.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:1900]: Same problem here (ish) as below - we should pass a Reference (which `$failureTransportsServiceLocator` is), not a Definition.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:1918]: we could revert this change to lessen the diff

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php
Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php:1884]: Should we be checking for `!isset($senderReferences[$transport['failure_transport']])` here instead? That would be consistent with the above code. It's more confusing, but there is the edge case that the failure_transport is not a registered transport, but just a "sender service id"

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (weaverryan) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:35]: Hmm. But could we make `$failureSenders` a FULL array of "receivers' => $failureSenderService.

Take this pseudo-config:

```yml
failure_transport: failure1
transports:
    async1: null
    async2:
       failure_transport: failure2
    failure1: null
    failure2: null
```

In this case, $failureSenders would be a map containing this:

```
[
    async1 => <failure1 sender service>
    async2 => <failure2 sender service>
    failure1 => <failure1 sender service>
    failure2 => <failure2 sender service>
]
```

Then we would have all the info we need in the map, correct? This "occurred" to me by looking *only* at this class (and not thinking about the rest of the code): in this class, it *feels* like we should be able to pass this object all the info it needs with one, full service locator of failure senders.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php
Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php:145]: If I'm correct in `FrameworkExtension`, these won't be needed.

File: src/Symfony/Component/Messenger/Command/FailedMessagesRetryCommand.php
Comment (weaverryan) [inline src/Symfony/Component/Messenger/Command/FailedMessagesRetryCommand.php:45]: Could we eliminate the new `$failureTransports` argument and instead remove the type-hint from `?ReceiverInterface $globalReceiver`? The 2nd arg would then be able to receive a `ReceiverInterface` OR a `ServiceLocator` and we would check the type and act accordingly. Maybe we would just pass that straight to the parent constructor, since it already has the deprecation.

File: src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php
Comment (weaverryan) [inline src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php:312]: Hmm, wait a second. This is subtle. But currently, in `FrameworkExtension`, you are populating arg 0 with `$senderReferences[$config['failure_transport']]`. But before, we were populating them with *receiver* references. Even for me, this makes my head spin a bit :P. You could technically have a failure transport "receiver" service but no "sender" service... though this is (at the very least) an edge case. Usually a sender & receiver are a "pair" because you've defined them as a transport. But this doesn't need to be the case.

The situation would be this: I create a `ReceiverInterface` service and tag it with `messenger.receiver` and `alias` `my_failure_receiver`. But, I do not actually register this as a transport. Then, at the command line, I use `messenger:failure:show --failure-transport=my_failure_receiver`. That is technically legal. And it's even possible (though unlikely) to set up this kind of situation if you're using AMQP (e.g. you actually set up a true `failure_transport` called `failure1`, but then with custom binding & routing rules, those messages ultimately end up in a queue connected with the custom `my_failure_receiver` receiver.

So.... that's why this code was originally here and I think it probably still needs to be. I don't think that's a problem, other than some of the new code in `FrameworkExtension` should actually live here.

File: src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php
Comment (monteiro) [reply src/Symfony/Component/Messenger/EventListener/SendFailedMessageToFailureTransportListener.php:35]: How do you which one is the global one? Because your map just says that one specific transport is assigned to one failure sender service. 

Specify a specific key for global? 

[
    "messenger_global" => <failure1 sender service>
    async1 => <failure1 sender service>
    async2 => <failure2 sender service>
    failure1 => <failure1 sender service>
    failure2 => <failure2 sender service>
]

**Comment by fabpot** [comment]
We've just moved away from `master` as the main branch to use `5.x` instead. Unfortunately, I cannot reopen the PR and change the target branch to `5.x`. Can you open a new PR referencing this one to not loose the discussion? Thank you for your understanding and for your help.

**Comment by monteiro** [comment]
opened new PR https://github.com/symfony/symfony/pull/38468 with 5.x as base

**Comment by luigi370** [comment]
This is just merged into symfony 5 right? is there any change to make it work on symfony 4.4 without updating the framework? thanks!

**Comment by stof** [comment]
updating the framework...
Old versions of Symfony don't get the new features of newer versions (that would defeat the purpose of versioning)

**Pull request closed without merging**
